bin/Macronify/Main.cpp

```diff
@@ -0,0 +1,734 @@
+        std::cerr << maybe_jobs.TakeError() << std::endl;
+        return EXIT_FAILURE;
+    }
+
+    for (const auto &job : maybe_jobs.TakeValue()) {
+        auto maybe_ast = job.Run();
+        if (!maybe_ast.Succeeded()) {
+            std::cerr << maybe_ast.TakeError() << std::endl;
+            return EXIT_FAILURE;
+        }
+
+        ast.emplace(maybe_ast.TakeValue());
+
+        mlir::DialectRegistry registry;
+
+        // Register the MLIR dialects we will be converting to
+        registry.insert<
+            vast::hl::HighLevelDialect,
+            macroni::macroni::MacroniDialect,
+            macroni::kernel::KernelDialect
+        >();
+        mctx.emplace(registry);
+        macroni::MacroniMetaGenerator meta(*ast, &*mctx);
+        vast::cg::CodeGenBase<macroni::Visitor> codegen(&*mctx, meta);
+        builder.emplace(&*mctx);
+
+        codegen.append_to_module(ast->UnderlyingAST().getTranslationUnitDecl());
+        mlir::OwningOpRef<mlir::ModuleOp> mod = codegen.freeze();
+
+        // TODO(bpp): Add a command-line argument to convert special macros into
+        // special operations
+
+        // Register conversions
+
+        // Mark expansions of get_user(), offsetof() and container_of() as
+        // illegal
+        mlir::ConversionTarget trg(*mctx);
+        trg.addDynamicallyLegalOp<macroni::macroni::MacroExpansion>(
+            [](Op *op) {
+                using ME = macroni::macroni::MacroExpansion;
+                ME exp = mlir::dyn_cast<ME>(op);
+                bool has_results = !exp.getResultTypes().empty();
+                llvm::StringRef macro_name = exp.getMacroName();
+                size_t num_params = exp.getParameterNames().size();
+                bool is_get_user = (has_results &&
+                                    num_params == 2 &&
+                                    macro_name == get_user),
+                    is_offsetof = (has_results &&
+                                   num_params == 2 &&
+                                   macro_name == offsetof),
+                    is_container_of = (has_results &&
+                                       num_params == 3 &&
+                                       macro_name == container_of),
+                    is_rcu_dereference = (has_results &&
+                                          num_params == 1 &&
+                                          macro_name == rcu_dereference),
+                    //   NOTE(bpp): The actual smp_mb() macro may not have
+                    //   results! This check will need to be updated if so.
+                    is_smp_mb = (has_results &&
+                                 num_params == 0 &&
+                                 macro_name == smp_mb);
+
+                using MP = macroni::macroni::MacroParameter;
+                using RMO = vast::hl::RecordMemberOp;
+                using PT = vast::hl::PointerType;
+                using TA = mlir::TypeAttr;
+                if (is_get_user) {
```

Comment:
Each of these should be separate fns. Then this can reduce to
```
if ( is_X( op ) )
  return is_X_legal( op );
if ( is_Y( op ) )
  ...
```
or something similar. I would wrap all of these legality queries in a class scope.